docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+selected candidate peer.
+If this is the case, the peer occupying this connection slot is set to the
+[upgrading state](#upgrading-peer), so that it will be evicted once the
+connection to the candidate peer is successfully established.
+
+### Dialing peer
+
+A peer that has been returned to the router as the next peer to dial.
+The router should be attempting to connect to this peer.
+
+A peer in `Dialing` state it is not considered as a candidate peer.
+
+### Dialed transition
+
+This transition is performed when the node establishes an outgoing connection
+with a peer.
+This means that the peer manager has provided this peer to the router as the
+[next peer to dial](#dialnext-transition), and the router has dialed and
+successfully established a connection with the peer.
+The peer is thus expected to be in the [`Dialing`](#dialing-peer) state.
+
+It may occur, however, that when this transition is invoked the peer is already
+in the [`Connected`](#connected-peer) state.
+The most likely reason is that the router, while dialing this peer, has also
+accepted an incoming connection from the same peer.
+In this case, the transition fails, indicating to the router that is should
+close the newly established connection.
+
+> Question: is it possible to have multiple routines dialing to the same peer?
+> This could be another reason for this transition to fail.
+
+It may also occur that the node is already connected to `MaxConnected` peers,
+which means that all connection slots are full.
+In this case, the peer manager tries to find a connection slot that can be
+[upgraded](#slots-upgrading) to give room for the new established connection.
+If no suitable connection slot is found, or the hard limit `MaxConnected +
+MaxConnectedUpgrade` of connected peers is reached, the transitions fails.
+
+Notice that, in order to dial this peer, the peer manager may have put another
+lower-ranked peer in the [upgrading sub-state](#upgrading-peeer) to give room
+to this connection.
+In this case, the slot for the established connection was *reserved*, and this
+transition will not fail.
+
+If the transition succeeds, the peer is set to the
+[`Connected`](#connected-peer) state as an `outgoing` peer.
+The peer's `LastConnected` and the dialed address' `LastDialSuccess` times are
+set, and dialed address' `DialFailures` counter is reset.
+
+> If the peer is `Inactive`, it is set as active.
+> This action has no effect apart from producing metrics.
+
+If a connection slot was upgraded to give room for the established connection, the
+peer on that slot transitions to the [evict sub-state](#evict-peer).
+
+#### Errors
+
+The transition fails if:
+
+- the node dialed itself
+- the peer is already in the `Connected` state
+- the node is connected to `MaxConnected` peers, and no slot is suitable for upgrading
+
+Errors are also returned if:
+
+- the dialed peer was removed from the peer store
```

Comment:
This was not commented in this document, probably it should have.

There is a maximum number of peers we keep in the peer store, this is a parameter. When this size is reached, we prune some peers. We prune low-ranked peers, and try to avoid pruning peers we set to some state (like Dialing), but in some corner cases we can indeed prune them.